packages/calypso-e2e/src/lib/pages/media-page.ts

```diff
@@ -150,4 +158,45 @@ export class MediaPage extends BaseContainer {
 		await cancelButton.click();
 		await this.page.waitForSelector( selectors.mediaModal );
 	}
+
+	/**
+	 * Uploads the file to the Media gallery.
+	 *
+	 * @param {string} fullPath Full path to the file on disk.
+	 * @returns {Promise<void>} No return value.
+	 */
+	async upload( fullPath: string ): Promise< void > {
+		await this.page.setInputFiles( selectors.addNewButton, fullPath );
+	}
+
+	/**
+	 * Confirms the media file upload was successful.
+	 *
+	 * @param {string} fullPath Full path on disk.
+	 * @returns {Promise<void>} No return value.
+	 */
+	async confirmUploadSuccessful( fullPath: string ): Promise< void > {
+		// Extract the basename and extension of the file from the full path.
+		const basename = path.basename( fullPath );
+		const gallery = await this.page.waitForSelector( selectors.gallery );
+		await gallery.waitForElementState( 'stable' );
+		// Each media item has the attribute `title=<filename.extension>` and Playwright
+		// is able to select elements by attribute.
+		await gallery.waitForSelector( `[title="${ basename }"]` );
+	}
```

Comment:
I think we're mixing concerns here. It would be expected for the `upload` function to throw if the upload failed (as a self-validation, in case it's used for any other test), so in the test we'd just need to check if that error happened. What if the `upload` function was sth like this:

```js
async upload( filepath ) {
	await this.page.setInputFiles( selectors.addNewButton, filepath );

	const filename = path.basename( filepath );
	const itemSelector = `figure[title="${ filename }"]`;
	const result = await Promise.race( [
		this.page.waitForSelector( itemSelector ),
		this.page.waitForSelector( selectors.uploadRejectionNotice ),
	] );

	if ( ( await result.getAttribute( 'title' ) ) === filename ) {
		return result;
	}

	const noticeText = await result.$( '.notice__text' ).innerText();
	throw new Error( noticeText );
}
```

And then in the test:

```js
it( 'Upload a permitted file type and see it is added to the gallery' ) {
	const uploadedItem = await mediaPage.upload( filepath );
	assert( await uploadedItem.isVisible() );
}

it( 'Upload a forbidden file type and see the rejection notice' ) {
	try {
		await mediaPage.upload( filepath );
	} catch ( error ) {
		assert.match( error.message, /file type is not supported/ );
	}	
}
```

What do you think?

Reply:
I think you are totally right here. 

It's been 2 weeks since I wrote this and I am unsure why I decided to make methods for success and rejection. The wonders of spending some time away from the code.

I've revised the test and the MediaPage.upload() method along the lines of what you have suggested. For the spec itself, I have decided to parametrize the `it` step.